README.md
![](https://img.shields.io/chrome-web-store/v/gnhhdgbaldcilmgcpfddgdbkhjohddkj.svg) ![](https://img.shields.io/chrome-web-store/users/gnhhdgbaldcilmgcpfddgdbkhjohddkj.svg) ![](https://img.shields.io/chrome-web-store/rating/gnhhdgbaldcilmgcpfddgdbkhjohddkj.svg) ![](https://img.shields.io/chrome-web-store/stars/gnhhdgbaldcilmgcpfddgdbkhjohddkj.svg) ![](https://img.shields.io/chrome-web-store/rating-count/gnhhdgbaldcilmgcpfddgdbkhjohddkj.svg) 

![NiM Logo](https://june07.github.io/image/smallPromoTile.png)

![Google Reviews](https://june07.github.io/image/312uiu.gif)
# NiM (Node.js --inspector Manager)
###### Streamlines your development process

* [Chrome Web Store](http://bit.ly/2W8hQG9) (works with any Chromium browsers: [Google's Chrome](https://www.google.com/chrome/), [Microsoft's Edge](https://www.microsoftedgeinsider.com/en-us/, ), [Opera](https://www.opera.com/), [Vivaldi](https://vivaldi.com/), [Brave](https://brave.com/), [Epic](https://www.epicbrowser.com/), and more...  )


## Blog Posts/Updates:
Now integrates with VSCode using NiMS [VSCode Extension](http://bit.ly/2VwKeSZ) 
![VSCode Integration](https://imgur.com/download/PRMn9md)
* Why I wrote NiM https://blog.june07.com/inspect-broke-my-workflow/
* Custom DevTools feature: https://blog.june07.com/nim-custom-devtools-url/
## Easier than `about::inspect (chrome://inspect/#devices)`

NiM manages the Chrome DevTools window/tab lifecycle leaving you with more ability to focus on what matters... debugging your code.  You no longer need to copy/paste DevTools URL's or continue opening/closing tabs/windows.

NiM automatically detects the URL that is generated when running node (locally or remotely) with --inspect option. NiM provides you with the option of automatically opening and closing Chrome DevTools in a tab or window. Just toggle the Manual/Auto setting and then start a debugging session.  DevTools will open either on clicking the "Open DevTools" button or after the specified timeout period.  If set to auto close, once you end your debugging session, The DevTools tab/window will close automatically.
 
 * Manage and monitor local and remote debugging sessions
 * Manual or automatic control of DevTools interface
 * Open DevTools in a new tab or window
 * Make DevTools focused or inactive on start
 * Customize duration between v8 Inspector probes
 * Autosave settings
 * Debug node processes launched by [VSCode](https://imgur.com/download/PRMn9md)


## Setup / Usage / How To

Simple and basic... just install the Chrome Extension in any of the following ways:

1. Install via Chrome Web Store:
    https://bit.ly/2W8hQG9 (https://chrome.google.com/webstore/detail/gnhhdgbaldcilmgcpfddgdbkhjohddkj)
2. Or, download .crx file directly from releases as shown here:

![https://media.giphy.com/media/xT0xenBpYPF2F0j2fe/giphy.gif](https://media.giphy.com/media/xT0xenBpYPF2F0j2fe/giphy.gif)



# Need Additional Information?
* NiM post install page provides some help [https://blog.june07.com/nim-install](https://blog.june07.com/nim-install/)
* Debugging NiM itself [https://blog.june07.com/debugging-nim/](https://blog.june07.com/debugging-nim/)

*Note: At the time of writing, the v8 --inspect option is fairly new. See https://nodejs.org/api/debugger.html#debugger_v8_inspector_integration_for_node_js for additional details on the option.*
#### If you enjoy using NIM please [give us a 5 star rating and/or a G+1.](https://chrome.google.com/webstore/detail/nim-node-inspector-manage/gnhhdgbaldcilmgcpfddgdbkhjohddkj/reviewshttps://chrome.google.com/webstore/detail/nim-node-inspector-manage/gnhhdgbaldcilmgcpfddgdbkhjohddkj/reviews)

Any and all feedback is encouraged and welcome.  [Send us an email!](mailto:667@june07.com)

[![IPv4 Lease](https://june07.github.io/image/EVERYTHING.jpg)](https://june07.com/blog/ipv4/?utm_source=NiM&utm_medium=options%20page&utm_campaign=ipv4&utm_content=2)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
_Added, Changed, Deprecated, Removed, Fixed, Security_
## [2.2.0] - 2019-05-25
### Added
- Added setting to have inspector window maximized.  https://github.com/june07/NiM/issues/70

## [2.1.1] - 2019-05-23
### Added
- Added devToolsCompat setting and made default as most recent version of Chrome seems to have the missing file browser again.

## [2.1.0] - 2019-05-13
### Added
- Added VSCode Integration https://marketplace.visualstudio.com/items?itemName=June07.nims
### Updated
- Updated deps
### Removed
- Removed Google Plus support as it's deprecated.  https://developers.google.com/+/api-shutdown

## [2.0.2] - 2019-03-28
### Fixed
- Attempt to fix https://github.com/june07/NiM/pull/69.  Will need to wait before coming to a solid conclusion as the problem is very sporadic in nature and not always reproducible.

## [2.0.1] - 2019-03-11
### Fixed
- Runaway tabs when port fowarding

## [2.0.0] - 2019-01-16
### Added
-Tabs!  Multiple simultaneous debug sessions are now possible with tabs.
-Tab notification to make identifying the active debugging tab easier

## [1.4.0] - 2019-01-14
### Added
- Added encryption
### Updated
- Improved locking code
## [1.3.2] - 2019-01-14
### Added
- Added option to have 'panel' style window.  Thank you @T99Rots for the initial PR https://github.com/june07/NiM/pull/64
## [1.3.1] - 2018-12-30
### Removed
- Removed emails from analytics.
## [1.3.0] - 2018-10-22
### Added
- Added 8 more langulage translations for Bengali, Bulgarian, Hindi, Macedonian, Malay, Persian, Swahil, and Tamil.  Thank you @aaliyahw for https://github.com/june07/NiM/pull/58, https://github.com/june07/NiM/pull/59, https://github.com/june07/NiM/pull/61, and https://github.com/june07/NiM/pull/62
## [1.2.1] - 2018-10-18
### Fixed
- Fixed a bug causing locks to pile up.
## [1.2.0] - 2018-10-14
### Updated
- Updated appspot DevTools version to Chrome 71.
### Added
- Added june07 hosted DevTools version.  Added blog post to describe the custom DevTools option https://june07.com/blog/nim-custom-devtools-url.
## [1.1.0] - 2018-10-08
### Updated
- i18N updates to 31 languages.
## [1.0.0] - 2018-10-04
### Added
- Feature to select which version of DevTools to use including a custom version.  This feature stemmed from the following Chrome bug https://bugs.chromium.org/p/chromium/issues/detail?id=851853, and will also work as a fix.
- DevTools panel for future NiM capabilities.
### Updated
- UI Improvements including moving options from a popup to a page.
## [0.15.1] - 2018-06-07
### Updated
- Added update_url per https://productforums.google.com/forum/?hl=en#!topic/chrome/kGgLwnrDKpQ;context-place=forum/chrome in hopes to fix https://github.com/june07/NiM/issues/34
## [0.15.0] - 2018-04-26
### Updated
- Polish languge updates.  Thank you @kubeek https://github.com/june07/NiM/pull/44
### Fixed
- Breaking change in Chrome Version 66.0.3359.117 (Official Build) (64-bit) see https://github.com/june07/NiM/issues/47 and https://github.com/june07/NiM/issues/46 for more info.
## [0.14.5] - 2017-11-28
### Updated
- Changed default setting back to "auto" .
### Added
- Added install page for  timely messages (i.e.  a possible solution to that nasty crashing bug), etc.
## [0.14.4] - 2017-11-16
Updated
- Changed default settings to "manual" to see if that somehow helps the crashing bug.
## [0.14.3] - 2017-11-06
### Updated
- Lots of deps.
### Changed
- Changed default key bindings for open devtools shortcut.
## [0.14.2] - 2017-10-26
### Fixed
- Forgot to update locales during GitHub Issue #32fix  push.
## [0.14.1] - 2017-10-23
### Changed
- Renamed "Hostname" label on popup to "Host" to more accurately represent the data field.  Also added/changed the regexp to match and validate data.
## [0.14.0] - 2017-10-16
### Added
- Notification upon using the keyboard shortcut.  The extra permission requirement is so that noticiations can be displayed.
### Fixed
- Bug where the auto/manual setting would revert upon enable/disable of the extension.
## [0.13.12] - 2017-10-08
### Fixed
- Regression error from yesterdays push.  Regex should not have started and ended from beginnnig and ending of line respectively.
## [0.13.9] - 2017-10-07
### Changed
- Improve Indonesian localization https://github.com/june07/Node.js-inspector-Manager-NiM-/pull/25,  and improved remote debugging host https://github.com/june07/Node.js-inspector-Manager-NiM-/pull/26.  Thank you to fatkhanfauzi for both PRs!
- Improve Russian localization https://github.com/june07/Node.js-inspector-Manager-NiM-/pull/22, Thank you to artofvs!
## [0.13.8] - 2017-05-23
### Fixed
- "Make Window Focused" option not working.  Error was due to what I believe is a bug in the Chrome code base...?  Anyhow this works around the dependancy.
## [0.13.7] - 2017-05-22
### Fixed
- Reload loop in some cases.
## [0.13.6] - 2017-05-21
### Fixed
- Stuck in a locked state in some cases.
## [0.13.5] - 2017-05-08
### Fixed
- Pretty major bug related to locking fixed.
## [0.13.4] - 2017-04-22
### Fixed
- Minor fixed to logging.
## [0.13.3] - 2017-04-17
### Added
- Faster polling (GitHub - Allow users to pick faster polling rate #13)
- Better debugging options for NiM itself using DevTools instead of the popup window.  See https://june07.com/nim/debug for details.
### Fixed
- Bug where settings were not being saved upon restart of Chrome.
## [0.12.5] - 2017-03-10
### Fixed
- Not a bug per se but I've received a lot of feedback about it so I added some alerting under certain situations.  One being when the user is trying to connect to a non-existant DevTools instance, and/or one on the wrong host, port, etc.
## [0.12.4] - 2017-03-06
### Fixed
- GitHub Issue #12 breaking change caused by 0.12.2.
## [0.12.2] - 2017-03-05
### Fixed

popup.js
/**
 * MIT License
 *
 *    Copyright (c) 2016-2019 June07
 *
 *    Permission is hereby granted, free of charge, to any person obtaining a copy
 *    of this software and associated documentation files (the "Software"), to deal
 *    in the Software without restriction, including without limitation the rights
 *    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *    copies of the Software, and to permit persons to whom the Software is
 *    furnished to do so, subject to the following conditions:
 *
 *    The above copyright notice and this permission notice shall be included in all
 *    copies or substantial portions of the Software.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 *    SOFTWARE.
*/
var ngApp = angular.module('NimPopupApp', ['angularMoment']);
ngApp
.config([
  '$compileProvider',
  function($compileProvider) {   
    $compileProvider.imgSrcSanitizationWhitelist(/^\s*(https?|ftp|mailto|chrome-extension):/);
    $compileProvider.aHrefSanitizationWhitelist(/^\s*(https?|ftp|mailto|chrome-extension):/);
  } 
])
.filter('stringLimit', ['$filter', function($filter) {
   return function(input, limit, wordBoundry) {
      if (! input) return;
      if (input.length <= limit && limit > 0) {
          return input;
      }
      var limitedString = $filter('limitTo')(input, Math.abs(limit));
      if (limit < 0)
        return input.substring(limitedString.lastIndexOf(" ")+1, input.length)
      if (wordBoundry)
        return limitedString.substring(0, limitedString.lastIndexOf(" "))
      return limitedString + '...';
   };
}])
.filter('nodeProgram', function() {
    return (cmd) => {
      let parts = cmd.split(' ');
      return parts[parts.length-1];
    };
})
.controller('nimPopupController', ['$scope', '$window', function ($scope, $window) {
    $scope.sortType = 'date';
    $scope.sortReverse = true;
    $scope.active = "none";
    var bgWindow = $window.chrome.extension.getBackgroundPage();
    var controllerElement = bgWindow.document.querySelector('body');
    $scope.bg = bgWindow.angular.element(controllerElement).scope()
    $scope.bg.localize($window, function() {});
    $scope.yieldResult = "wait";
    $scope.messageModalState = "closed";

    var  chrome = $window.chrome,
      $ = $window.$;
    
    $scope.openModal = function() {
      $.notify.close();
      $scope.pn.next("wait");
    }
    $scope.clickHandler = function () {
        $scope.bg.save("host");
        $scope.bg.save("port");
        $scope.bg.openTab($scope.bg.settings.host, $scope.bg.settings.port, function (error, result) {
            if (error && typeof error === "string") {
              showErrorMessage(error);
              if (error === 'DevTools is already open.') $scope.bg.tabNotification({host: $scope.bg.settings.host, port: $scope.bg.settings.port});
            }
            if (error) showErrorMessage(error.statusText);
            else
              $scope.message = result;
        });
    };
    $scope.clickHandlerOpenLocalDevTools = function(translatedURL) {
      let localHost = translatedURL.split(/ws=(.*)\//)[1].split(':')[0];
      let localPort = translatedURL.split(/ws=(.*)\//)[1].split(':')[1];
      $scope.bg.openTab(localHost, localPort, function (error, result) {
        if (error && typeof error === "string") {
          showErrorMessage(error);
          if (error === 'DevTools is already open.') $scope.bg.tabNotification({host: localHost, port: localPort});
        }
        if (error && typeof error === "object" && error.message) showErrorMessage(error.message);
        if (error) showErrorMessage(error.statusText);
        else $scope.message = result;
      });
    }
    $scope.clickHandlerRemoveLocalDevToolsSession = function(sessionID) {
      $scope.bg.removeLocalSession(sessionID)
    }
    $scope.switchHandler = function() {
        $scope.bg.save("auto");
    }
    $scope.track = function (url) {
        $window._gaq.push(['_trackPageview', url]);
    };
    function showErrorMessage(error) {
      $window.document.querySelector('#site-href').style.display = "none";
      $window.Materialize.toast(error, 5000);
      var siteHrefTimeout;
      if (siteHrefTimeout) clearTimeout(siteHrefTimeout);
      siteHrefTimeout = setTimeout(function() {
        $window.document.querySelector('#site-href').style.display = "inline";
      }, 5050)
    }
    function trackInputClick(e) {
        $window._gaq.push(['_trackEvent', e.target.id, 'clicked']);
    }
    var userInputs = $window.document.getElementsByClassName('ga-track');

    function trackInputClickListener(event) {
        trackInputClick(event);
    }
    function keypressHandler(event) {
        if (event.keyCode === 13) {
            event.preventDefault();
            $scope.clickHandler();
            $window._gaq.push(['_trackEvent', 'User Event', 'OpenDevTools', 'Enter Key Pressed', '', true]);
        }
    }
    for (var i = 0; i < userInputs.length; i++) {
      userInputs[i].addEventListener('click', trackInputClickListener);
      if (userInputs[i].id === "port" || userInputs[i].id === "hostname")
        userInputs[i].addEventListener('keypress', keypressHandler);
    }
    $window.document.querySelector('#modal1 > div.modal-header > button').addEventListener('click', function() {
      $('#modal1').modal('close');
    });
    $window.document.querySelector('#modal2 > div.modal-header > button').addEventListener('click', function() {
      $('#modal2').modal('close');
    });
    $window.document.querySelector('#modal3 > div.modal-header > button').addEventListener('click', function() {
      $('#modal3').modal('close');
    });
    $window.document.querySelector('#modal4 > div.modal-header > button.close').addEventListener('click', function() {
      $('#modal4').modal('close');
    });
    $window.document.querySelector('#options-button').addEventListener('click', function() {
      if (chrome.runtime.openOptionsPage) {
        // New way to open options pages, if supported (Chrome 42+).
        chrome.runtime.openOptionsPage();
      } else {
        // Reasonable fallback.
        $window.open(chrome.runtime.getURL('options.html'));
      }
    });
    $('.modal').modal({
      dismissible: true, // Modal can be dismissed by clicking outside of the modal
      opacity: .5, // Opacity of modal background
      in_duration: 300, // Transition in duration
      out_duration: 200, // Transition out duration
      starting_top: '4%', // Starting top style attribute
      ending_top: '10%', // Ending top style attribute
      ready: function() {
        $scope.messageModalState = "open";
        if ($scope.notify) $scope.notify.close();
      }
    });
    $('#modal1').perfectScrollbar();
    $('#modal2').perfectScrollbar();
    $scope.openOptionsPage = function() {
      if (chrome.runtime.openOptionsPage) {
        // New way to open options pages, if supported (Chrome 42+).
        chrome.runtime.openOptionsPage();
      } else {
        // Reasonable fallback.
        $window.open(chrome.runtime.getURL('options.html'));
      }
    };
    $($window.document).ready(() => {
      if ($scope.bg.settings.nimsVscode.enabled) $scope.bg.NiMSVSCodeConnector.check();
      $('ul.tabs').tabs({ onShow: (tab) => {
        $scope.bg.updateLocalSessions();
        $scope.bg.state.popup.selectedTab = tab[0].id;
        $scope.$apply();
        $scope.initTippyTips();
        $scope.initConnectionErrorMessage();
      }});
      $window.tippy.setDefaults({
        allowHTML: true,
        theme: 'nim',
        placement: 'bottom',
        size: 'large',
        interactive: false,
        hideOnClick: 'toggle'
      })
      $scope.initTippyTips();
      $scope.initConnectionErrorMessage();
    });
    $scope.initTippyTips = function() {
      $scope.tippyTips = $window.tippy('.tippy');
      $scope.tippyTips.forEach((tip) => {
        let el = document.querySelector('[id^=' + tip.props.content.substring(1) + ']');
        switch (tip.props.content) {
          case '#tippyTemplateGithubNodejsNodeIssues24085':
            tip.set({
              trigger: 'click',
              theme: 'githubNodejsNodeIssues24085',
              placement: 'top',
              size: 'large'
            });
          break;
        }
        tip.setContent(el);
      });
    }
    $scope.showTippy = function(event) {
      let parent = event.currentTarget;
      let tip = parent._tippy;
      tip.set({interactive: true});
      tip.popper.addEventListener("mouseleave", (event) => {
        event.currentTarget._tippy.hide();
        event.currentTarget._tippy.set({interactive: false});
      });
    }
    $scope.initConnectionErrorMessage = function() {
      $('.ml11').each((i, el) => {
        if (!el.classList.contains('pretty')) {
          $(el).addClass('pretty');
          $('.ml11 .letters').each(function(){
            $(this).html($(this).text().replace(/([^\x00-\x80]|\w)/g, "<span class='letter'>$&</span>").replace(/(^\| )/, "<blink class='carat'>$&</blink>"));
          });
          $window.anime.timeline({loop: true})
          .add({
            targets: '.ml11 .line',
            scaleY: [0,1],
            opacity: [0.5,1],
            easing: "easeOutExpo",
            duration: 700
          })
          .add({
            targets: '.ml11 .line',
            translateX: [0,$(".ml11 .letters").width()],
            easing: "easeOutExpo",
            duration: 700,
            delay: 100
          }).add({
            targets: '.ml11 .letter',
            opacity: [0,1],
            easing: "easeOutExpo",
            duration: 600,
            offset: '-=775',
            delay: function(el, i) {
              return 34 * (i+1)
            }
          }).add({
            targets: '.ml11',
            opacity: 0,
            duration: 1000,
            easing: "easeOutExpo",
            delay: 1000
          });
        }
      });
    }
}]);